app/main.py

```diff
@@ -1,2 +1,128 @@
+from __future__ import annotations
+from typing import Any, Iterable, Hashable
+
+
+_not_provided = object()
+
+
 class Dictionary:
-    pass
+    DEFAULT_SIZE = 8
+    THRESHOLD = 2 / 3
+
+    def __init__(self, items: Iterable = (), **kwargs) -> None:
+        self.length = 0
+        self.hash_table = [None] * self.DEFAULT_SIZE
+
+        for key, value in items:
+            self[key] = value
+        self.update(kwargs)
+
+    def __getitem__(self, key: Hashable) -> Any:
+        _, _, value = self.hash_table[self._get_with_hash(key, hash(key))]
+        return value
+
+    def __setitem__(self, key: Hashable, value: Any) -> None:
+        key_hash = hash(key)
+
+        if self.length + 1 > len(self.hash_table) * self.THRESHOLD:
+            self._resize()
+
+        self._set_with_hash(key, key_hash, value)
+
+    def __delitem__(self, key: Hashable) -> None:
+        self.hash_table[self._get_with_hash(key, hash(key))] = None
+        self.length -= 1
+
+    def _resize(self) -> None:
+        old_hash_table = self.hash_table
+
+        self.length = 0
+        self.hash_table = [None] * len(old_hash_table) * 2
+
+        for saved_key, saved_hash, saved_value in filter(None, old_hash_table):
+            self._set_with_hash(saved_key, saved_hash, saved_value)
+
+    def _get_with_hash(self, key: Hashable, hash_: int) -> int:
+        hash_table_len = len(self.hash_table)
+        first_index = index = hash_ % hash_table_len
+
+        while True:
```

Comment:
you can separate logic of finding index in a separate method and then reuse it in both set and get methods